src/main/java/de/filefighter/rest/domain/filesystem/business/FileSystemBusinessService.java

```diff
@@ -96,11 +94,45 @@ public void deleteFileSystemItemById(long fsItemId, User authenticatedUser) {
             throw new FileSystemItemCouldNotBeDeletedException(fsItemId);
 
         if (fileSystemEntity.isFile()) {
-            // CASE 0 - File
             fileSystemRepository.delete(fileSystemEntity);
         } else {
-            // CASE 1 - Folder
-
+            HashSet<FileSystemEntity> foundFolders = new HashSet<>();
+            foundFolders.add(fileSystemEntity);
+            Iterator<FileSystemEntity> iterator = foundFolders.iterator();
+
+            // Lists for change or deletion
+            ArrayList<FileSystemEntity> entitiesToBeDeleted = new ArrayList<>();
+            ArrayList<FileSystemEntity> entitiesToBeChanged = new ArrayList<>();
+
+            while (iterator.hasNext()) {
```

Comment:
1. Multithreading is not a option?

2. What about an recursive approach? First find all the folders, then go from the bottom up and delete what can be deleted.

I don't know if this handles the case when a subfolder can be deleted, but the main folder is kept, because it contains a not editable file.

Reply:
1. Multithreading sounds like a nightmare because iterators are not threadsafe.
2. I already thought about reading in the folders in a `TreeSet` and afterwards deleting it. That sounds like a trade of (possible) stability for more redundancy, because i have to walk the whole filetree to find the folders.

Reply:
We could discuss this. I also thought about creating a own `FileSystem` and `Path` instance to use premade algos for recursive deleting. (like `walk`)